File: SubscriberClient/src/subscriber/SubscriberBase.java
package subscriber;

import com.pcbsys.nirvana.client.nAsyncExceptionListener;
import com.pcbsys.nirvana.client.nBaseClientException;
import com.pcbsys.nirvana.client.nEventAttributes;
import com.pcbsys.nirvana.client.nEventProperties;
import com.pcbsys.nirvana.client.nIllegalArgumentException;
import com.pcbsys.nirvana.client.nRealmUnreachableException;
import com.pcbsys.nirvana.client.nReconnectHandler;
import com.pcbsys.nirvana.client.nSecurityException;
import com.pcbsys.nirvana.client.nSession;
import com.pcbsys.nirvana.client.nSessionAlreadyInitialisedException;
import com.pcbsys.nirvana.client.nSessionAttributes;
import com.pcbsys.nirvana.client.nSessionFactory;
import com.pcbsys.nirvana.client.nSessionNotConnectedException;
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.Properties;
import java.util.StringTokenizer;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Base class that contains standard functionality for a nirvana sample app
 */
public abstract class SubscriberBase implements nReconnectHandler, nAsyncExceptionListener {

	protected static final String RNAME = "RNAME";
	protected static final String LOGLEVEL = "LOGLEVEL";
	protected static final String HPROXY = "HPROXY";
	protected static final String HAUTH = "HAUTH";
	protected static final String UM_USERNAME = "UM_USERNAME";
	protected static final String UM_PASSWORD = "UM_PASSWORD";
	protected static final String PASS_PASSWORD_IN_CONSOLE = "PASS_PASSWORD_IN_CONSOLE";

	protected nSession mySession = null;
	protected nSessionAttributes nsa = null;
	private String myLastSessionID;

	private static final String DEFAULT_USERNAME = "JavaApplication";

	protected static Properties getConfigProperties() {

		Properties prop = new Properties();
		try (InputStream input = new FileInputStream("resources/config.properties")) {
			prop.load(input);
			// System.out.println(prop.getProperty("RNAME"));

		} catch (IOException ex) {
			ex.printStackTrace();
		}

		return prop;
	}

	/**
	 * This method processes a string consisting of one or more comma separated
	 * RNAME values and splits them into a a String[]
	 *
	 * @param realmdetails The RNAME of the Nirvana realm
	 */
	protected static String[] parseRealmProperties(String realmdetails) {
		if (realmdetails.contains("(")) {
			Pattern hSPattern = Pattern.compile("\\((.*?)\\)");
			Matcher m = hSPattern.matcher(realmdetails);
			ArrayList<String> result = new ArrayList<String>();
			while (m.find()) {
				result.add(m.group(1));
			}
			String[] tmp = new String[result.size()];
			for (int i = 0; i < result.size(); i++) {
				tmp[i] = "(" + result.get(i) + ")";
			}
			return tmp;
		}

		ArrayList<String> result = new ArrayList<String>();

		StringTokenizer st = new StringTokenizer(realmdetails, ",");
		while (st.hasMoreTokens()) {
			result.add(st.nextToken());
		}

		return result.toArray(new String[result.size()]);
	}

	/**
	 * This method demonstrates the Nirvana API calls necessary to construct a
	 * nirvana session
	 *
	 * @param realmDetails a String[] containing the possible RNAME values
	 */
	protected void constructSession(String[] realmDetails, Properties configProperties) {

		// Create a realm session attributes object from the array of strings
		try {
			nsa = new nSessionAttributes(realmDetails, 2);
			nsa.setFollowTheMaster(false);
			nsa.setName(getClass().getSimpleName());
		} catch (Exception ex) {
			System.out.println("Error creating Session Attributes. Please check your RNAME");
			System.exit(1);
		}

		String username = configProperties.getProperty(UM_USERNAME);
		String password = configProperties.getProperty(UM_PASSWORD);
		// Add this class as an asynchronous exception listener
		try {
			// Create a session object from the session attributes object, passing this
			// as a reconnect handler class (optional). This will ensure that the
			// reconnection
			// methods will get called by the API.

			if (password != null) {
				mySession = nSessionFactory.create(nsa, this, username, password);
			} else {
				mySession = nSessionFactory.create(nsa, this, username);
			}
			mySession.addAsyncExceptionListener(this);
			mySession.enableThreading(4);
		} catch (nIllegalArgumentException ex) {
			//
		}

		// Initialise the Nirvana session. This physically opens the connection to the
		// Nirvana realm, using the specified protocols. If multiple interfaces are
		// supported
		// these will be attempted in weight order (SSL, HTTPS, socket, HTTP).
		try {
			mySession.init();
			myLastSessionID = mySession.getId();
		}
		// Handle errors
		catch (nSecurityException sec) {
			System.out.println("The current user is not authorised to connect to the specified Realm Server");
			System.out.println("Please check the realm acls or contact support");
			sec.printStackTrace();
			System.exit(1);
		} catch (nRealmUnreachableException rue) {
			System.out.println("The Nirvana Realm specified by the RNAME value is not reachable.");
			System.out.println("Please ensure the Realm is running and check your RNAME value.");
			rue.printStackTrace();
			System.exit(1);
		} catch (nSessionNotConnectedException snce) {
			System.out.println("The session object used is not physically connected to the Nirvana Realm.");
			System.out.println("Please ensure the Realm is up and check your RNAME value.");
			snce.printStackTrace();
			System.exit(1);
		} catch (nSessionAlreadyInitialisedException ex) {
			System.out.println("The session object has already been initialised.");
			System.out.println("Please make only one call to the .init() function.");
			ex.printStackTrace();
			System.exit(1);
		}
	}

	/**
	 * A callback is received by the API to this method to notify the user of a
	 * disconnection from the realm. The method is enforced by the nReconnectHandler
	 * interface but is normally optional. It gives the user a chance to log the
	 * disconnection or do something else about it.
	 *
	 * @param anSession The Nirvana session being disconnected
	 */
	public void disconnected(nSession anSession) {
		try {
			System.out.println("You have been disconnected from " + myLastSessionID);
		} catch (Exception ex) {
			System.out.println("Error while disconnecting " + ex.getMessage());
		}
	}

	/**
	 * A callback is received by the API to this method to notify the user of a
	 * successful reconnection to the realm. The method is enforced by the
	 * nReconnectHandler interface but is normally optional. It gives the user a
	 * chance to log the reconnection or do something else about it.
	 *
	 * @param anSession The Nirvana session being reconnected
	 */
	public void reconnected(nSession anSession) {
		try {
			myLastSessionID = mySession.getId();
			System.out.println("You have been Reconnected to " + myLastSessionID);
		} catch (Exception ex) {
			System.out.println("Error while reconnecting " + ex.getMessage());
		}
	}

	/**
	 * A callback is received by the API to this method to notify the user that the
	 * API is about to attempt reconnecting to the realm. The method is enforced by
	 * the nReconnectHandler interface but is normally optional. It allows the user
	 * to decide whether further attempts are required or not, whether custom delays
	 * should be enforced etc.
	 *
	 * @param anSession The Nirvana session that will be used to reconnect
	 */
	public boolean tryAgain(nSession anSession) {
		try {
			System.out.println("Attempting to reconnect to " + System.getProperties().get("RNAME"));
		} catch (Exception ex) {
			System.out.println("Error while trying to reconnect " + ex.getMessage());
		}
		return true;
	}

	/**
	 * A callback is received by the API to this method to notify the user that the
	 * an asynchronous exception (in a thread different than the current one) has
	 * occured.
	 *
	 * @param ex The asynchronous exception that was thrown
	 */
	public void handleException(nBaseClientException ex) {
		System.out.println("An Asynchronous Exception was received from the Nirvana realm.");
		ex.printStackTrace();
	}

	protected void displayEventProperties(nEventProperties prop) {
		System.out.println("----------------------------------------------------------------");
		displayEventProperties(prop, 0);
		System.out.println("----------------------------------------------------------------");
	}

	protected void displayEventProperties(nEventProperties prop, int level) {
		String tab = "";
		for (int x = 0; x < level; x++) {
			tab += "\t";
		}
		System.out.println(tab + "Event Prop : ");
		List<String> list = Collections.<String>list(prop.getKeysAsStrings());
		Collections.sort(list, new Comparator<String>() {
			@Override
			public int compare(String o1, String o2) {
				return o1.compareTo(o2);
			}
		});
		// nEventPropertiesIterator keys = prop.getKeyIterator();
		Iterator keys = list.iterator();
		while (keys.hasNext()) {
			Object key = keys.next();
			Object value = prop.get(key.toString());
			if (value instanceof nEventProperties) {
				nEventProperties pvalue = (nEventProperties) value;
				System.out.println(tab + "[" + key + "(event prop)]:");
				displayEventProperties(pvalue, level + 1);
			} else if (value instanceof nEventProperties[]) {
				nEventProperties[] pvalue = (nEventProperties[]) value;
				System.out.println(tab + "[" + key + "(event prop[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					displayEventProperties(pvalue[x], level + 1);
				}
			} else if (value instanceof String[]) {
				String[] pvalue = (String[]) value;
				System.out.println(tab + "[" + key + "(String[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println("   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof long[]) {
				long[] pvalue = (long[]) value;
				System.out.println(tab + "[" + key + "(long[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println("   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof int[]) {
				int[] pvalue = (int[]) value;
				System.out.println(tab + "[" + key + "(int[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println(tab + "   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof byte[]) {
				byte[] pvalue = (byte[]) value;
				System.out.println(tab + "[" + key + "(byte[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println(tab + "   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof boolean[]) {
				boolean[] pvalue = (boolean[]) value;
				System.out.println(tab + "[" + key + "(boolean[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println(tab + "   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof double[]) {
				double[] pvalue = (double[]) value;
				System.out.println(tab + "[" + key + "(double[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println(tab + "   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof short[]) {
				short[] pvalue = (short[]) value;
				System.out.println(tab + "[" + key + "(short[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println(tab + "   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else if (value instanceof char[]) {
				char[] pvalue = (char[]) value;
				System.out.println(tab + "[" + key + "(char[])]:");
				for (int x = 0; x < pvalue.length; x++) {
					System.out.println(tab + "   [" + key + "]:[" + x + "]=" + pvalue[x]);
				}
			} else {
				System.out.println(tab + key.toString() + " = " + value.toString());
			}
		}
	}

	protected void displayEventAttributes(nEventAttributes attributes) {
		System.out.println("Merge Allowed : " + attributes.allowMerging());
		System.out.println("Message Type  : " + attributes.getMessageType());
		System.out.println("Delivery mode : " + attributes.getDeliveryMode());
		System.out.println("Priority      : " + attributes.getDeliveryMode());
		if (attributes.getApplicationId() != null) {
			System.out.println("Application Id : " + new String(attributes.getApplicationId()));
		}
		if (attributes.getCorrelationId() != null) {
			System.out.println("Correlation Id : " + new String(attributes.getCorrelationId()));
		}
		if (attributes.getMessageId() != null) {
			System.out.println("Message Id     : " + new String(attributes.getMessageId()));
		}
		if (attributes.getPublisherHost() != null) {
			System.out.println("Published from : " + new String(attributes.getPublisherHost()));
		}
		if (attributes.getPublisherName() != null) {
			System.out.println("Published by   : " + new String(attributes.getPublisherName()));
		}
		if (attributes.getTimestamp() != 0) {
			System.out.println("Published on   : " + new Date(attributes.getTimestamp()).toString());
		}
		if (attributes.getReplyToName() != null) {
			System.out.println("Reply To       : " + new String(attributes.getReplyToName()));
		}
		if (attributes.getExpiration() != 0) {
			System.out.println("Expires on     : " + attributes.getExpiration());
		}
		if (attributes.isRedelivered()) {
			System.out.println("Redelivered event : " + attributes.getRedeliveredCount());
		}
	}
} // End of subscriber Class

File: UniversalMessaging/UniversalMessagingClient/SubscriberBase.cs
using System;
using System.Collections.Generic;
using com.pcbsys.nirvana.client;

namespace UniversalMessagingClient
{

    /**
     * Base class that contains standard functionality for a nirvana sample app
     */
    public class SubscriberBase : nReconnectHandler, nAsyncExceptionListener
    {

        protected nSession mySession = null;
        protected nSessionAttributes nsa = null;
        private String myLastSessionID;
        private static System.Text.UTF8Encoding encoding = new System.Text.UTF8Encoding();

        protected virtual void processArgs(String[] args) { }

        protected static void processEnvironmentVariables()
        {
            //Process Environment Variables
        }

        /**
         * This method processes a string consisting of one or more comma separated
         * RNAME values and splits them into a a String[]
         *
         * @param realmdetails The RNAME of the Nirvana realm
         */
        protected static String[] parseRealmProperties(String realmdetails)
        {
            String[] rproperties = new String[4];
            int idx = 0;
            String[] st = realmdetails.Split(',');
            for (int x = 0; x < st.Length; x++)
            {
                String someRNAME = (String)st[x];
                rproperties[idx] = someRNAME;
                idx++;
            }
            //Trim the array
            String[] rpropertiesTrimmed = new String[idx];
            System.Array.Copy(rproperties, 0, rpropertiesTrimmed, 0, idx);
            return rpropertiesTrimmed;
        }

        /**
         * This method demonstrates the Nirvana API calls necessary to construct a
         * nirvana session
         *
         * @param realmDetails a String[] containing the possible RNAME values
         */
        protected void constructSession(String[] realmDetails)
        {

            //Create a realm session attributes object from the array of strings
            try
            {
                nsa = new nSessionAttributes(realmDetails, 2);
            }
            catch (Exception)
            {
                Console.WriteLine(
                    "Error creating Session Attributes. Please check your RNAME");
            }

            //Add this class as an asynchronous exception listener
            try
            {
                //Create a session object from the session attributes object, passing this
                //as a reconnect handler class (optional). This will ensure that the reconnection
                //methods will get called by the API.
                mySession = nSessionFactory.create(nsa, this);
                mySession.addAsyncExceptionListener(this);
            }
            catch (nIllegalArgumentException) { }

            //Initialise the Nirvana session. This physically opens the connection to the
            //Nirvana realm, using the specified protocols. If multiple interfaces are supported
            //these will be attempted in weight order (SSL, HTTPS, socket, HTTP).
            try
            {
                nConstants.setClientLogLevel(1);


                mySession.init();


                myLastSessionID = mySession.getId();
                //mySession.updateConnectionListWithServerList();
            }
            //Handle errors
            catch (nSecurityException sec)
            {
                Console.WriteLine("The current user is not authorised to connect to the specified Realm Server");
                Console.WriteLine("Please check the realm acls or contact support");
                Console.WriteLine(sec.StackTrace);
                Environment.Exit(1);
            }
            catch (nRealmUnreachableException)
            {
                Console.WriteLine(
                    "The Nirvana Realm specified by the RNAME value is not reachable.");
                Console.WriteLine(
                    "Please ensure the Realm is running and check your RNAME value.");
                Environment.Exit(1);
            }
            catch (nSessionNotConnectedException)
            {
                Console.WriteLine(
                    "The session object used is not physically connected to the Nirvana Realm.");
                Console.WriteLine(
                    "Please ensure the Realm is up and check your RNAME value.");
                Environment.Exit(1);
            }
            catch (nSessionAlreadyInitialisedException)
            {
                Console.WriteLine("The session object has already been initialised.");
                Console.WriteLine("Please make only one call to the .init() function.");
                Environment.Exit(1);
            }
        }

        /**
         * A callback is received by the API to this method to notify the user of a disconnection
         * from the realm. The method is enforced by the nReconnectHandler interface but is normally optional.
         * It gives the user a chance to log the disconnection or do something else about it.
         *
         * @param anSession The Nirvana session being disconnected
         */
        public void disconnected(nSession anSession)
        {
            try
            {
                Console.WriteLine("You have been disconnected from " + myLastSessionID);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error while disconnecting " + ex.Message);
            }
        }

        /**
         * A callback is received by the API to this method to notify the user of a successful reconnection
         * to the realm. The method is enforced by the nReconnectHandler interface but is normally optional.
         * It gives the user a chance to log the reconnection or do something else about it.
         *
         * @param anSession The Nirvana session being reconnected
         */
        public void reconnected(nSession anSession)
        {
            try
            {
                myLastSessionID = mySession.getId();
                Console.WriteLine("You have been Reconnected to " + myLastSessionID);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error while reconnecting " + ex.Message);
            }
        }

        /**
         * A callback is received by the API to this method to notify the user that the API is about
         * to attempt reconnecting to the realm. The method is enforced by the nReconnectHandler
         * interface but is normally optional. It allows the user to decide whether further
         * attempts are required or not, whether custom delays should be enforced etc.
         *
         * @param anSession The Nirvana session that will be used to reconnect
         */
        public bool tryAgain(nSession anSession)
        {
            try
            {
                Console.WriteLine("Attempting to reconnect to ");
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error while trying to reconnect " + ex.Message);
            }
            return true;
        }

        /**
         * A callback is received by the API to this method to notify the user that the an
         * asynchronous exception (in a thread different than the current one) has occured.
         *
         * @param ex The asynchronous exception that was thrown
         */
        public void handleException(nBaseClientException ex)
        {
            Console.WriteLine("An Asynchronous Exception was received from the Nirvana realm.");
        }

        protected static void UsageEnv()
        {
            Console.WriteLine("\n\n(Environment Variables) \n");

            Console.WriteLine("(RNAME) - One or more RNAME entries in the form protocol://host:port");
            Console.WriteLine("   protocol - Can be one of nsp, nhp, nsps, or nhps, where:");
            Console.WriteLine("   nsp - Specifies Nirvana Socket Protocol (nsp)");
            Console.WriteLine("   nhp - Specifies Nirvana HTTP Protocol (nhp)");
            Console.WriteLine("   nsps - Specifies Nirvana Socket Protocol Secure (nsps), i.e. using SSL/TLS");
            Console.WriteLine("   nhps - Specifies Nirvana HTTP Protocol Secure (nhps), i.e. using SSL/TLS");
            Console.WriteLine("   port - The port number of the server");
            Console.WriteLine("\nHint: - For multiple RNAME entries, use comma separated values which will be attempted in connection weight order\n");
            Console.WriteLine("(LOGLEVEL) - This determines how much information the nirvana api will output 0 = verbose 7 = quiet\n");
            Environment.Exit(1);
        }

        protected void displayEventProperties(nEventProperties prop)
        {
            Console.WriteLine("Event Prop : ");
            IEnumerator<String> keys = prop.getKeys();
            keys.Reset();
            while (keys.MoveNext())
            {
                Object key = keys.Current;
                Object value = prop.get(key.ToString());
                if (value is nEventProperties)
                {
                    nEventProperties pvalue = (nEventProperties)value;
                    Console.WriteLine("[" + key + "(event prop)]:");
                    displayEventProperties(pvalue);
                }
                else if (value is nEventProperties[])
                {
                    nEventProperties[] pvalue = (nEventProperties[])value;
                    Console.WriteLine("[" + key + "(event prop[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        displayEventProperties(pvalue[x]);
                    }
                }
                else if (value is String[])
                {
                    String[] pvalue = (String[])value;
                    Console.WriteLine("[" + key + "(String[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        Console.WriteLine("   [" + key + "]:[" + x + "]=" + pvalue[x]);
                    }
                }
                else if (value is long[])
                {
                    long[] pvalue = (long[])value;
                    Console.WriteLine("[" + key + "(long[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        outputKeyVal(key, pvalue[x], x);
                    }
                }
                else if (value is int[])
                {
                    int[] pvalue = (int[])value;
                    Console.WriteLine("[" + key + "(int[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        outputKeyVal(key, pvalue[x], x);
                    }
                }
                else if (value is byte[])
                {
                    byte[] pvalue = (byte[])value;
                    Console.WriteLine("[" + key + "(byte[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        outputKeyVal(key, pvalue[x], x);
                    }
                }
                else if (value is bool[])
                {
                    bool[] pvalue = (bool[])value;
                    Console.WriteLine("[" + key + "(bool[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        outputKeyVal(key, pvalue[x], x);
                    }
                }
                else if (value is double[])
                {
                    double[] pvalue = (double[])value;
                    Console.WriteLine("[" + key + "(double[])]:");
                    for (int x = 0; x < pvalue.Length; x++)
                    {
                        outputKeyVal(key, pvalue[x], x);
                    }
                }
                else
                {
                    outputKeyVal(key, value);
                }
            }
        }

        private void outputKeyVal(object key, object val)
        {
            outputKeyVal(key, val, -1);
        }
        private void outputKeyVal(object key, object val, int idx)
        {
            if (idx == -1)
            {
                Console.WriteLine("   [" + key + "]\t=\t" + val);
            }
            else
            {
                Console.WriteLine("   [" + key + "]:[" + idx + "]\t=\t" + val);
            }
        }

        protected void displayEventAttributes(nEventAttributes attributes)
        {
            if (attributes.getApplicationId() != null)
                Console.WriteLine("Application Id : " + encoding.GetString(attributes.getApplicationId()));
            if (attributes.getCorrelationId() != null)
                Console.WriteLine("Correlation Id : " + encoding.GetString(attributes.getCorrelationId()));
            if (attributes.getMessageId() != null)
                Console.WriteLine("Message Id     : " + encoding.GetString(attributes.getMessageId()));
            if (attributes.getPublisherHost() != null)
                Console.WriteLine("Published from : " + encoding.GetString(attributes.getPublisherHost()));
            if (attributes.getPublisherName() != null)
                Console.WriteLine("Published by   : " + encoding.GetString(attributes.getPublisherName()));
            DateTime date = Jan1st1970.AddMilliseconds(attributes.getTimestamp());
            if (attributes.getTimestamp() != 0)
                Console.WriteLine("Published on   : " + date.ToString());
        }

        private static readonly DateTime Jan1st1970 = new DateTime
            (1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        public static long CurrentTimeMillis()
        {
            return (long)(DateTime.UtcNow - Jan1st1970).TotalMilliseconds;
        }

        //public static void Main(String[] args)
        //{
        //    SubscriberBase subscriberBase = new SubscriberBase();
        //    subscriberBase.constructSession(args[0].Split(','));
        //    subscriberBase.mySession.init();
        //}
    }
}

